### PR TITLE
Remove token on validation

### PR DESCRIPTION
**Pull request by akrabat**
Ensure token is removed in persisent mode on validation

When the token is successfully validated and we are not in persistent mode, then we need to remove the token. This is a separate operation to handling failure.

The side-effect of the previous code is that is allowed a replay attack where the same token could be used multiple times even when persisent mode was disabled.

Thanks to @xhlika for the report.

### CODE REVIEW

File: tests/GuardTest.php [hunk 1]
```diff
@@ -279,43 +279,35 @@ public function testEnforceStorageLimitWithIterator()
...
 
-        $mw->process($requestProphecy->reveal(), $requestHandlerProphecy->reveal());
-        $this->assertArrayNotHasKey('test_name123', $storage);
+        $request = $this->createMock(ServerRequestInterface::class);
+        $request->expects(self::once())->method('getMethod')->willReturn('POST');
+        $request->expects(self::exactly(2))->method('withAttribute')->willReturn($request);
+        $request->expects(self::once())->method('getParsedBody')->willReturn([
+           'test_name' => 'test_name',
+           'test_value' => 'test_value123',
+        ]);
```
Comment (l0gicgate) [inline tests/GuardTest.php:304]: Can we use prophecies instead of mocks for cohesion as all the other tests are written that way

File: .travis.yml [hunk 2]
```diff
@@ -19,8 +19,8 @@ before_script:
   - composer install -n
 
 script:
-  - if [[ "$ANALYSIS" != 'true' ]]; then vendor/bin/phpunit ; fi
-  - if [[ "$ANALYSIS" == 'true' ]]; then vendor/bin/phpunit --coverage-clover clover.xml ; fi
+  - if [[ "$ANALYSIS" != 'true' ]]; then vendor/bin/phpunit -dxdebug.mode=coverage; fi
+  - if [[ "$ANALYSIS" == 'true' ]]; then vendor/bin/phpunit -dxdebug.mode=coverage --coverage-clover clover.xml ; fi
```
Comment (l0gicgate) [inline .travis.yml:23]: Build is failing on `7.3` because of this 🤔

**Review by l0gicgate — CHANGES_REQUESTED** [review]

File: tests/GuardTest.php (hunk 1 above)
Comment (akrabat) [reply tests/GuardTest.php:304]: Sure. I didn't as PHPUnit deprecated Prophecy in [issue 4141](https://github.com/sebastianbergmann/phpunit/issues/4141), so it seems possible that we'll look to move to stock PHPUnit mocks over time.

Maybe that's a separate PR to re-write the lot though.

File: .travis.yml (hunk 2 above)
Comment (akrabat) [reply .travis.yml]: It's related to Xdebug 3.0. Turns out that `XDEBUG_MODE=coverage ./vendor/bin/phpunit …` is the correct format!

Comment (akrabat) [reply .travis.yml]: Rebased out incorrect versions.

File: tests/GuardTest.php (hunk 1 above)
Comment (l0gicgate) [reply tests/GuardTest.php:304]: I don't think rewriting all the tests would be time well spent, there's a shim that addresses this deprecation which we've implemented on all the other repos. We will need to add it here as well!

Comment (akrabat) [reply tests/GuardTest.php]: Unit test now uses Prophecy

File: tests/GuardTest.php [hunk 3]
```diff
@@ -279,43 +279,45 @@ public function testEnforceStorageLimitWithIterator()
...
             'test_name' => 'test_value123',
         ];
+
+        $responseProphecy = $this->prophesize(ResponseInterface::class)
+            ->willImplement(ResponseInterface::class);
+
+        $requestHandlerProphecy = $this->prophesize(RequestHandlerInterface::class);
+        $requestHandlerProphecy
+            ->handle(Argument::type(ServerRequestInterface::class))
+            ->willReturn($responseProphecy->reveal());
```
Comment (l0gicgate) [inline tests/GuardTest.php:292]: should add `->shouldBeCalledOnce()` so we get an assertion there that this is being called.

File: tests/GuardTest.php [hunk 4]
```diff
@@ -279,43 +279,45 @@ public function testEnforceStorageLimitWithIterator()
...
+                'test_name' => 'test_name',
+                'test_value' => 'test_value123',
             ])
             ->shouldBeCalledOnce();
 
-        $requestHandlerProphecy = $this->prophesize(RequestHandlerInterface::class);
 
         $mw->process($requestProphecy->reveal(), $requestHandlerProphecy->reveal());
-        $this->assertArrayNotHasKey('test_name123', $storage);
+        self::assertArrayNotHasKey('test_name', $storage);
```
Comment (l0gicgate) [inline tests/GuardTest.php:317]: I'm curious why is that changed to `self` instead of `$this`?

**Review by l0gicgate — CHANGES_REQUESTED** [review]

File: tests/GuardTest.php (hunk 4 above)
Comment (akrabat) [reply tests/GuardTest.php:317]: `assertArrayNotHasKey` is declared `static` in PHPUnit, so it _should_ be called with `self::` as it doesn't use any instance properties or methods. PHP allows you to call static methods with `$this->` though, so either works.

File: tests/GuardTest.php (hunk 3 above)
Comment (akrabat) [reply tests/GuardTest.php]: Fixed.

File: tests/GuardTest.php (hunk 4 above)
Comment (l0gicgate) [reply tests/GuardTest.php:318]: @akrabat today I learned 😄

**Review by l0gicgate — APPROVED** [review]
